odo/backends/sql.py

```diff
@@ -307,19 +317,34 @@ def dshape_to_alchemy(dshape):
     if isinstance(dshape, str):
         dshape = datashape.dshape(dshape)
     if isinstance(dshape, Option):
-        return dshape_to_alchemy(dshape.ty)
+        return dshape_to_alchemy(dshape.ty, dialect, level=level + 1)
     if str(dshape) in types:
         return types[str(dshape)]
     if isinstance(dshape, datashape.Record):
-        return [sa.Column(name,
-                          dshape_to_alchemy(typ),
-                          nullable=isinstance(typ[0], Option))
-                for name, typ in dshape.parameters[0]]
+        if 0 <= level <= 1:  # we only allow at most a single level of nesting
+            return [sa.Column(name,
+                              dshape_to_alchemy(typ, dialect, level=level + 1),
+                              nullable=isinstance(typ, Option))
+                    for name, typ in dshape.parameters[0]]
+        raise TypeError(
+            'nested record dshape not supported by sqlalchemy')
     if isinstance(dshape, datashape.DataShape):
         if isdimension(dshape[0]):
-            return dshape_to_alchemy(dshape[1])
+            if isscalar(dshape[1]):
+                type_ = array_types.get(dialect.name, None)
+                if type_ is None:
+                    raise TypeError('dialect %s has no array type for dshape '
+                                    '%s' % (dialect.name, dshape))
+                return type_(dshape_to_alchemy(dshape[1],
+                                               dialect, level=level + 1))
+            else:
+                if not level:
+                    return dshape_to_alchemy(dshape[1], dialect,
+                                             level=level + 1)
+                raise TypeError('nested record dshape not supported by sqlalchemy')
```

Comment:
Yes, thanks for pointing that out. I'll merge this into that and see what happens.